opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -2555,6 +2573,119 @@ namespace detail {
+        }
+        else
+#endif
+        {
+            if ( global_number_wells )
+            {
+                rate_converter_.defineState(reservoir_state);
+            }
+        }
+
+        std::vector<double> well_rates(np, 0.0);
+        std::vector<double> convert_coeff(np, 1.0);
+
+
+        if ( !well_voidage_rates.empty() ) {
+            for (int w = 0; w < nw; ++w) {
+                const bool is_producer = wells->type[w] == PRODUCER;
+
+                // not sure necessary to change all the value to be positive
+                if (is_producer) {
+                    std::transform(well_state.wellRates().begin() + np * w,
+                                   well_state.wellRates().begin() + np * (w + 1),
+                                   well_rates.begin(), std::negate<double>());
+
+                    // the average hydrocarbon conditions of the whole field will be used
+                    const int fipreg = 0; // Not considering FIP for the moment.
+
+                    rate_converter_.calcCoeff(well_rates, fipreg, convert_coeff);
+                    well_voidage_rates[w] = std::inner_product(well_rates.begin(), well_rates.end(),
+                                                               convert_coeff.begin(), 0.0);
+                } else {
+                    // TODO: Not sure whether will encounter situation with all zero rates
+                    // and whether it will cause problem here.
+                    std::copy(well_state.wellRates().begin() + np * w,
+                              well_state.wellRates().begin() + np * (w + 1),
+                              well_rates.begin());
+                    // the average hydrocarbon conditions of the whole field will be used
+                    const int fipreg = 0; // Not considering FIP for the moment.
+                    rate_converter_.calcCoeff(well_rates, fipreg, convert_coeff);
+                    std::copy(convert_coeff.begin(), convert_coeff.end(),
+                              voidage_conversion_coeffs.begin() + np * w);
+                }
+            }
+        }
+    }
+
+
+
+
+
+    template <class Grid, class WellModel, class Implementation>
+    void
+    BlackoilModelBase<Grid, WellModel, Implementation>::
+    applyVREPGroupControl(const ReservoirState& reservoir_state,
+                          WellState& well_state)
+    {
+        if (asImpl().wellModel().wellCollection()->havingVREPGroups()) {
+            std::vector<double> well_voidage_rates;
+            std::vector<double> voidage_conversion_coeffs;
+            computeWellVoidageRates(reservoir_state, well_state, well_voidage_rates, voidage_conversion_coeffs);
+            asImpl().wellModel().wellCollection()->applyVREPGroupControls(well_voidage_rates, voidage_conversion_coeffs);
+
+            // for the wells under group control, update the currentControls for the well_state
+            const size_t number_node = asImpl().wellModel().wellCollection()->numNode();
```

Comment:
Instead of using numNode() and getNode(), write this loop as a range-for loop over getLeafNodes() instead.